File: resource-registry/src/main/java/nl/haploid/octowight/registry/service/ResourceRegistryService.java
package nl.haploid.octowight.registry.service;

import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.AtomGroup;
import nl.haploid.octowight.registry.data.ResourceRoot;
import nl.haploid.octowight.registry.data.ResourceRootFactory;
import nl.haploid.octowight.registry.repository.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@Service
public class ResourceRegistryService {

	private Logger log = LoggerFactory.getLogger(getClass());

	@Autowired
	private ResourceRootDmoRepository resourceRootDmoRepository;

	@Autowired
	private ResourceElementDmoRepository resourceElementDmoRepository;

	@Autowired
	private ResourceRootFactory resourceRootFactory;

	@Autowired
	private ResourceRootDmoFactory resourceRootDmoFactory;

	@Autowired
	private VersionDmoRepository versionDmoRepository;

	public boolean isNewResource(final ResourceRoot resourceRoot) {
		final ResourceRootDmo dmo = resourceRootDmoRepository
				.findByResourceTypeAndAtomIdAndAtomTypeAndAtomLocus(resourceRoot.getResourceType(),
						resourceRoot.getAtomId(), resourceRoot.getAtomType(), resourceRoot.getAtomLocus());
		return dmo == null;
	}

	public ResourceRoot saveResource(final ResourceRoot resourceRoot) {
		final ResourceRootDmo resourceRootDmo = resourceRootDmoFactory.fromResourceRoot(resourceRoot);
		final ResourceRootDmo dmo = resourceRootDmoRepository.saveAndFlush(resourceRootDmo);
		log.debug(String.format("Saved resource: %s/%d", dmo.getResourceType(), dmo.getResourceId()));
		return resourceRootFactory.fromResourceDmo(dmo);
	}

	// TODO: test
	public List<ResourceRoot> markResourcesDirty(final AtomGroup atomGroup, final List<AtomChangeEvent> atomChangeEvents) {
		final List<Long> atomIds = atomChangeEvents.stream()
				.map(AtomChangeEvent::getAtomId)
				.collect(Collectors.toList());
		final List<ResourceElementDmo> resourceElementDmos = resourceElementDmoRepository
				.findByAtomIdInAndAtomTypeAndAtomLocus(atomIds, atomGroup.getAtomType(), atomGroup.getAtomLocus());
		return resourceElementDmos.stream()
				.map(this::getResourceRootDmo)
				.filter(Objects::nonNull)
				.map(this::markResourceDirty)
				.collect(Collectors.toList());
	}

	private ResourceRootDmo getResourceRootDmo(final ResourceElementDmo resourceElementDmo) {
		return resourceRootDmoRepository.findByResourceTypeAndResourceId(resourceElementDmo.getResourceType(), resourceElementDmo.getResourceId());
	}

	private ResourceRoot markResourceDirty(final ResourceRootDmo resourceRootDmo) {
		log.debug(String.format("Mark %s/%d as dirty", resourceRootDmo.getResourceType(), resourceRootDmo.getResourceId()));
		final VersionDmo versionDmo = versionDmoRepository.findNext();
		resourceRootDmo.setVersion(versionDmo.getVersion());
		return resourceRootFactory.fromResourceDmo(resourceRootDmoRepository.saveAndFlush(resourceRootDmo));
	}
}

File: resource-registry/src/main/scala/nl/haploid/octowight/registry/service/ResourceRegistryService.scala
package nl.haploid.octowight.registry.service

import nl.haploid.octowight.registry.data.{ResourceIdentifier, ResourceRoot}
import nl.haploid.octowight.registry.repository._
import nl.haploid.octowight.{AtomChangeEvent, AtomGroup}
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

import scala.collection.JavaConverters._

@Service
class ResourceRegistryService {
  private[this] lazy val log = LoggerFactory.getLogger(getClass)

  @Autowired private[this] val resourceRootDmoRepository: ResourceRootDmoRepository = null
  @Autowired private[this] val resourceElementDmoRepository: ResourceElementDmoRepository = null
  @Autowired private[this] val sequenceService: SequenceService = null

  def saveResource(resourceRoot: ResourceRoot): Option[ResourceRoot] = {
    Option(findResourceRootDmo(resourceRoot)) match {
      case Some(resourceRootDmo) => if (resourceRootDmo.tombstone) Some(untombstoneResource(resourceRootDmo)) else None
      case None => Some(saveNewResource(resourceRoot))
    }
  }

  def markResourcesDirty(atomGroup: AtomGroup, atomChangeEvents: Iterable[AtomChangeEvent]) = {
    val atomIds = atomChangeEvents.map(event => Long.box(event.id))
    val resourceElementDmos = resourceElementDmoRepository.findByAtomIdInAndAtomCategoryAndAtomOrigin(
      atomIds.asJava, atomGroup.category, atomGroup.origin)
    resourceElementDmos.asScala
      .flatMap(element => findResourceRootDmo(element))
      .map(markResourceDirty)
  }

  def findResourceRoots(resourceIdentifiersByCollection: Map[String, Iterable[ResourceIdentifier]]): Map[String, Iterable[ResourceRoot]] = {
    val resourceRoots: Iterable[ResourceRoot] = resourceIdentifiersByCollection flatMap {
      case (collection, identifiers) =>
        val ids = identifiers.map(identifier => Long.box(identifier.id))
        resourceRootDmoRepository.findByResourceCollectionAndResourceIdIn(collection, ids.toList.asJava).asScala
          .map(ResourceRoot(_))
    }
    resourceRoots.groupBy(_.resourceCollection)
  }

  private[this] def findResourceRootDmo(resourceRoot: ResourceRoot): ResourceRootDmo = {
    resourceRootDmoRepository.findByResourceCollectionAndRootIdAndRootCategoryAndRootOrigin(
      resourceRoot.resourceCollection, resourceRoot.root.id, resourceRoot.root.category, resourceRoot.root.origin)
  }

  private[this] def findResourceRootDmo(resourceElementDmo: ResourceElementDmo) = {
    Option(resourceRootDmoRepository.findByResourceCollectionAndResourceId(
      resourceElementDmo.resourceCollection, resourceElementDmo.resourceId))
  }

  private[this] def markResourceDirty(resourceRootDmo: ResourceRootDmo) = {
    log.info(s"Mark ${resourceRootDmo.resourceCollection}/${resourceRootDmo.resourceId} as dirty")
    val version = sequenceService.nextValue(ResourceRootDmo.VersionSequence)
    val resourceRootDmoToSave = resourceRootDmo.copy(version = version)
    ResourceRoot(resourceRootDmoRepository.save(resourceRootDmoToSave))
  }

  private[this] def saveNewResource(resourceRoot: ResourceRoot): ResourceRoot = {
    val resourceId: Long = sequenceService.nextValue(ResourceRootDmo.IdSequence)
    val version: Long = sequenceService.nextValue(ResourceRootDmo.VersionSequence)
    val resourceRootDmoToSave = ResourceRootDmo(resourceRoot).copy(resourceId = resourceId, version = version)
    log.info(s"Save resource ${resourceRootDmoToSave.resourceCollection}/${resourceRootDmoToSave.resourceId}")
    ResourceRoot(resourceRootDmoRepository.save(resourceRootDmoToSave))
  }

  private[this] def untombstoneResource(resourceRootDmo: ResourceRootDmo): ResourceRoot = {
    log.info(s"Remove tombstone for ${resourceRootDmo.resourceCollection}/${resourceRootDmo.resourceId}")
    val resourceRootDmoToSave = resourceRootDmo.copy(tombstone = false)
    ResourceRoot(resourceRootDmoRepository.save(resourceRootDmoToSave))
  }
}
